smart-contracts/for-eth-devs/gas-model.md

```diff
@@ -1 +1,85 @@
+
+The system continuously translates between these representations: converting weight to gas when estimating costs, and converting gas back to weight when executing transactions.
+
+## How Gas Estimation Works
+
+When your wallet requests a gas estimate (`eth_estimateGas`), the Polkadot Hub performs a dry-run of your transaction. This test execution discovers:
+
+- How much computational time the contract will consume (`ref_time`)
+- How much state data needs to be verified (`proof_size`)
+- Whether any storage deposits are required (`storage_deposit`)
+
+The system then calculates a gas estimate that covers all these costs, including:
+
+- Base transaction overhead
+- Transaction length fees (cost per byte)
+- The actual contract execution costs
+- Any storage deposits
+
+The gas estimate also includes a small safety buffer to account for slight differences between the test run and actual execution.
+
+## Dynamic Gas Pricing
+
+PolkaVM uses dynamic pricing through a "fee multiplier" that adjusts based on network congestion:
+
+- When blocks are full, the multiplier increases, making transactions more expensive
+- When blocks are empty, the multiplier decreases, making transactions cheaper
+- The multiplier updates after every block based on utilization
+
+This creates a market-based pricing mechanism similar to Ethereum's base fee, helping to manage network resources efficiently.
+
+The gas price returned during estimation is simply the current fee multiplier value.
+
+!!! warning "Important for Users"
+    Because the fee multiplier can change between when you estimate gas and when your transaction executes, you can add a safety buffer (10-20%) to both your gas limit and gas price. This ensures your transaction will execute successfully even if network conditions change slightly.
+
+## Transaction Execution Flow
+
+The following diagram illustrates the complete lifecycle of a transaction from submission to settlement:
+
+```mermaid
+graph TD
+    U[User/Wallet] --> M[Transaction pool]
+    M --> P[Pre-dispatch convert gas to weight and create hold]
+    P --> C{Sufficient funds}
+    C -->|No| R[Rejected]
+    C -->|Yes| X[Execute contract within limits]
+    X --> S[Settle fee from actual weight and length; refund]
+    S --> B[Included in block]
+```
+
+The transaction execution flow is as follows:
+
+- **Pool and pre-dispatch**: The transaction enters the pool, `gas` is mapped to `weight`, and a temporary hold is created for the maximum fee exposure.
+- **Funds check**: If the hold is insufficient, the transaction is rejected before any execution.
+- **Execution**: If sufficient, the contract runs within the derived weight limits; a `storage_deposit` may be reserved when new storage is created.
+- **Settlement**: Fees are charged from the actual `weight` used plus the length fee; any unused hold is refunded.
```

Comment:
Is `weight`: `proof_size` + `ref_time`?

Reply:
And what is "length fee"?

Reply:
I know you have this somewhat defined already on this page but can we make it a bit more clear or a bit more emphasized?